File: Problems 0700-0799/799. Champagne Tower/solution.cpp
class Solution {
public:
    double champagneTower(int poured, int query_row, int query_glass) {
        vector<double> currRow(1, poured);

        for (int row = 0; row <= query_row; row++) {
            vector<double> nextRow(row+2, 0);
            for (int i = 0; i <= row; i++) {
                if (currRow[i] >= 1) {
                    nextRow[i] += (currRow[i]-1)/2;
                    nextRow[i+1] += (currRow[i]-1)/2;
                    currRow[i] = 1;
                }
            }
            if (row != query_row) swap(currRow, nextRow);
        }

        return currRow[query_glass];
    }
};

File: Problems 0700-0799/799. Champagne Tower/solution.java
class Solution {
    public double champagneTower(int poured, int query_row, int query_glass) {
        // double[] prev = new double[query_row + 1];
        // prev[0] = poured;

        // for (int row = 0; row < query_row; row++) {
        //     double[] curr = new double[query_row + 1];
        //     for (int cup = 0; cup <= row; cup++) {
        //         double leftover = Math.max(0, prev[cup] - 1);
        //         double halved = leftover / 2;
        //         curr[cup] += halved;
        //         curr[cup + 1] += halved;
        //     }

        //     prev = curr;
        // }

        // return Math.min(1.0, prev[query_glass]);

        double[] dp = new double[query_row + 1];
        dp[0] = poured;

        for (int row = 0; row < query_row; row++) {
            for (int cup = row; cup >= 0; cup--) {
                if (dp[cup] > 1) {
                    double leftover = dp[cup] - 1;
                    double halved = leftover / 2;
                    dp[cup] = halved;
                    dp[cup + 1] += halved;
                } else {
                    dp[cup] = 0;
                }
            }
        }

        return Math.min(1.0, dp[query_glass]);
    }
}